sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala

```diff
@@ -117,12 +119,32 @@ class InMemoryTable(
         this
       }
 
-      override def buildForBatch(): BatchWrite = writer
+      override def withQueryId(queryId: String): WriteBuilder = {
+        assert(!queryIdProvided, "queryId provided twice")
```

Comment:
later on can we continue the work in https://github.com/apache/spark/pull/25990 ? It still has value to give stronger guarantees. So that implementations don't need to do check like this.